src/kadmin/ktutil/ktutil_funcs.c

```diff
@@ -165,9 +166,13 @@ krb5_error_code ktutil_add(context, list, princ_str, kvno,
                                     &password.length);
         if (retval)
             goto cleanup;
-        retval = krb5_principal2salt(context, princ, &salt);
-        if (retval)
-            goto cleanup;
+        if (salt_str) {
+            salt = string2data(salt_str);
```

Comment:
This leaves salt in the precarious position of containing an alias to salt_str in one case and allocated data in the other.  Right now salt is leaked, which is a spearate bug.  I will adjust the commit to address this problem (and probably fix the memory leak), since our conventions for dealing with this kind of problem aren't easy to explain in text.